106-list-unregister-nodes-admin-api.md

```diff
@@ -0,0 +1,93 @@
+For example, when brokers are initially deployed as mixed-mode nodes (controller + broker) and later have their broker role removed, they remain registered.
+Although these nodes become fenced and controller-only, they may still receive partition assignments for new topics.
+The current implementation does not detect or handle this scenario, leading to missing unregistration.
+This issue has been reported by the community in [GitHub issue #11477](https://github.com/strimzi/strimzi-kafka-operator/issues/11477).
+
+## Motivation
+
+With the introduction of [KIP-1073](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1073%3A+Return+fenced+brokers+in+DescribeCluster+response) in Apache Kafka 4.0.0, the Kafka Admin API now provides the ability to list all registered brokers, including fenced ones.
+This makes it possible to manage unregistration entirely through standard Kafka APIs.
+
+This removes the need for Strimzi to track nodes registration status manually and enables proper handling of edge cases like mixed-node role changes.
+As a result, it also allows for a proper resolution of issue [#11477](https://github.com/strimzi/strimzi-kafka-operator/issues/11477).
+
+## Proposal
+
+This proposal suggests removing the custom registration tracking and instead relying on Apache Kafka's built-in registration tracking available via the Admin API.
+
+### Key changes
+
+The key changes are:
+
+* deprecate the `status.registeredNodeIds` field from the `Kafka` custom resource. The field will remain in the CRD but the operator won't set it in the `Kafka` status.
+* use the Kafka Admin client's `describeCluster()` method (with the option of including the fenced nodes) to retrieve the list of currently registered nodes. This call is done within a newly added `KafkaNodeUnregistration.listRegisteredNodes` method.
+* compare the list of registered brokers to the current set of active nodes and determine which ones need to be unregistered.
+* unregister nodes using the existing `KafkaNodeUnregistration.unregisterNodes()` method.
+
+### Reconciliation flow
+
+During a reconciliation, the operator goes through the following steps:
+
+* get current nodes IDs (both brokers and controllers) from the `KafkaCluster.nodes()` as is done today. It also extracts which ones are controller only nodes (not in mixed mode).
+* call `KafkaNodeUnregistration.listRegisteredNodes` method using the Admin client connected to the brokers' bootstrap service to retrieve all registered brokers.
+* build a `nodeIdsToUnregister` list with the nodes IDs to be unregistered.
+* unregister the nodes via the `KafkaNodeUnregistration.unregisterNodes` method as is done today.
+
+The `nodeIdsToUnregister` list is created in the following way: 
+
+* define `previousNodeIds` as the set of all registered brokers returned by Kafka.
+* define `fencedControllerOnlyNodeIds` as the subset of fenced nodes that are now controller-only (i.e., no longer brokers from being mixed-node).
```

Comment:
this sounds a bit confusing to me because fence state only applies to brokers, but I can't think of better name. Maybe it doesn't matter that much.

Reply:
Yeah I understand. I guess we can explain that in a comment/Java doc when implementing it. I will resolve this comment since I couldn't come up with a suggestion.